serde_derive/src/de.rs

```diff
@@ -844,9 +854,46 @@ fn deserialize_adjacently_tagged_enum(
         };
     }
 
-    let visit_third_key = quote! {
-        // Visit the third key in the map, hopefully there isn't one.
-        match try!(_serde::de::MapAccess::next_key_seed(&mut __map, #tag_or_content)) {
+    /// Advance the map by one key, returning early in case of error.
+    let next_key = quote! {
+        try!(_serde::de::MapAccess::next_key_seed(&mut __map, #tag_or_content))
+    };
+
+    /// When allowing unknown fields, we want to transparently step through keys we don't care
+    /// about until we find `tag`, `content`, or run out of keys.
+    let next_relevant_key = if deny_unknown_fields {
+        next_key
+    } else {
+        quote! {
+            {
+                let mut __rk : Option<_serde::private::de::TagOrContentField> = None;
```

Comment:
Please use absolute paths for `Option` and `None` (below as well) to avoid bugs like https://github.com/serde-rs/serde/issues/737#issuecomment-276728419.